GVFS/GVFS.Common/ProductUpgraderInfo.cs

```diff
@@ -10,39 +10,59 @@ public class ProductUpgraderInfo
         public const string UpgradeDirectoryName = "GVFS.Upgrade";
         public const string LogDirectory = "Logs";
         public const string DownloadDirectory = "Downloads";
+        public const string HighestAvailableVersionFileName = "HighestAvailableVersion";
 
         protected const string RootDirectory = UpgradeDirectoryName;
 
-        private static readonly HashSet<string> GVFSInstallerFileNamePrefixCandidates = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        public static bool IsLocalUpgradeAvailable(ITracer tracer)
         {
-            "SetupGVFS",
-            "VFSForGit"
-        };
+            try
+            {
+                Version highestAvailableVersion = ReadHighestRecordedAvailableVersion();
+                if (highestAvailableVersion == null)
+                {
+                    return false;
+                }
 
-        public static bool IsLocalUpgradeAvailable(string installerExtension)
-        {
-            string downloadDirectory = GetAssetDownloadsPath();
-            if (Directory.Exists(downloadDirectory))
+                Version currentVersion = new Version(CurrentGVFSVersion());
+                return highestAvailableVersion > currentVersion;
+            }
+            catch (Exception ex) when (
+                ex is IOException ||
+                ex is UnauthorizedAccessException ||
+                ex is NotSupportedException)
             {
-                foreach (string file in Directory.EnumerateFiles(downloadDirectory, "*", SearchOption.TopDirectoryOnly))
+                if (tracer != null)
                 {
-                    string[] components = Path.GetFileName(file).Split('.');
-                    int length = components.Length;
-                    if (length >= 2 &&
-                        GVFSInstallerFileNamePrefixCandidates.Contains(components[0]) &&
-                        installerExtension.Equals(components[length - 1], StringComparison.OrdinalIgnoreCase))
-                    {
-                        return true;
-                    }
+                    tracer.RelatedError(
+                        CreateEventMetadata(ex),
+                        "Exception encountered when determining if an upgrade is available.");
                 }
             }
 
             return false;
         }
 
-        public static IEnumerable<string> PossibleGVFSInstallerNamePrefixes()
+        public static void RecordHighestAvailableVersion(Version highestAvailableVersion)

```

Comment:
Use `PhysicalFileSystem` rather than `System.IO` so that we can unit test this method